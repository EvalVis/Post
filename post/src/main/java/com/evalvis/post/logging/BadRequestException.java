package com.evalvis.post.logging;

public class BadRequestException extends RuntimeException {
    public BadRequestException(String message) {
        super(message);
    }
}
