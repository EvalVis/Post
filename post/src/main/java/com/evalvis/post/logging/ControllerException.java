package com.evalvis.post.logging;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

import java.util.UUID;

@ControllerAdvice
public class ControllerException {
    private static final Logger log = LoggerFactory.getLogger(ControllerException.class);

    @ExceptionHandler
    public ResponseEntity<String> clientException(BadRequestException e) {
        return clientError(e, HttpStatus.BAD_REQUEST);
    }

    @ExceptionHandler
    public ResponseEntity<String> clientException(RestNotFoundException e) {
        return clientError(e, HttpStatus.NOT_FOUND);
    }

    @ExceptionHandler
    public ResponseEntity<String> clientException(UnauthorizedException e) {
        return clientError(e, HttpStatus.UNAUTHORIZED);
    }

    @ExceptionHandler
    public ResponseEntity<String> badCredentialsException(BadCredentialsException e) {
        return clientError(e, "Bad credentials.", HttpStatus.UNAUTHORIZED);
    }

    private ResponseEntity<String> clientError(Exception e, HttpStatus status) {
        return clientError(e, e.getMessage(), status);
    }

    private ResponseEntity<String> clientError(Exception e, String responseMessage, HttpStatus status) {
        String errorCode = UUID.randomUUID().toString();
        log.info("Client error code: {}, ", errorCode, e);
        return new ResponseEntity<>(responseMessage + " Error code: " + errorCode, status);
    }

    @ExceptionHandler
    public ResponseEntity<String> exception(Exception e) {
        String errorCode = UUID.randomUUID().toString();
        log.error("Error code: {}, ", errorCode, e);
        return new ResponseEntity<>("Something went wrong. Error code: " + errorCode, HttpStatus.INTERNAL_SERVER_ERROR);
    }
}
