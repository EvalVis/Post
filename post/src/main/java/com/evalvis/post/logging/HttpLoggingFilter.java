package com.evalvis.post.logging;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.annotation.WebFilter;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;
import org.springframework.web.util.ContentCachingRequestWrapper;
import org.springframework.web.util.ContentCachingResponseWrapper;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.Collection;
import java.util.Enumeration;

@Component
@WebFilter(filterName = "HttpLoggingFilter")
public final class HttpLoggingFilter extends OncePerRequestFilter {
    private static final Logger log = LoggerFactory.getLogger(HttpLoggingFilter.class);

    @Value("${post.http-masking-turned-on}")
    private boolean isHttpMaskingOn;

    @Override
    protected void doFilterInternal(
            @NonNull HttpServletRequest request, @NonNull HttpServletResponse response,
            FilterChain filterChain
    ) throws ServletException, IOException {
        ContentCachingRequestWrapper requestWrapper = new ContentCachingRequestWrapper(request);
        ContentCachingResponseWrapper responseWrapper = new ContentCachingResponseWrapper(response);
        long startTime = System.currentTimeMillis();
        try {
            filterChain.doFilter(requestWrapper, responseWrapper);
        } catch (ServletException | IOException e) {
            log.warn(
                    "While processing this request, exception was caused: "
                            + requestLog(request, requestWrapper).toPrettyString()
            );
            throw e;
        }
        long timeTaken = System.currentTimeMillis() - startTime;
        ObjectNode root = requestLog(request, requestWrapper);
        String responseBody = new String(
                responseWrapper.getContentAsByteArray(), response.getCharacterEncoding()
        );
        ObjectNode responseNode = root.withObject("/response");
        responseNode
                .put("status", response.getStatus())
                .put("body", isHttpMaskingOn ? "[hidden]" : responseBody)
                .put("time taken (ms)", timeTaken);

        appendResponseHeaders(responseNode, response);
        log.info(root.toPrettyString());
        responseWrapper.copyBodyToResponse();
    }

    private ObjectNode requestLog(
            HttpServletRequest request, ContentCachingRequestWrapper requestWrapper
    ) throws UnsupportedEncodingException {
        String requestBody = new String(
                requestWrapper.getContentAsByteArray(), request.getCharacterEncoding()
        );
        ObjectNode root = new ObjectMapper().createObjectNode().withObject("/HTTP");
        ObjectNode requestNode = root.withObject("/request");
        requestNode.put("URI", request.getRequestURI());
        requestNode.put("method", request.getMethod());
        requestNode.put("body", isHttpMaskingOn ? "[hidden]" : requestBody);
        appendRequestHeaders(requestNode, request);
        return root;
    }

    private void appendRequestHeaders(ObjectNode requestNode, HttpServletRequest request) {
        ArrayNode headers = requestNode.putArray("headers");
        Enumeration<String> headerNames = request.getHeaderNames();
        while(headerNames.hasMoreElements()) {
            String headerName = headerNames.nextElement();
            ObjectNode header = new ObjectMapper()
                    .createObjectNode()
                    .put(headerName, isHttpMaskingOn ? "[hidden]": request.getHeader(headerName));
            headers.add(header);
        }
    }

    private void appendResponseHeaders(ObjectNode responseNode, HttpServletResponse response) {
        ArrayNode headers = responseNode.putArray("headers");
        Collection<String> headerNames = response.getHeaderNames();
        for(String headerName : headerNames) {
            ObjectNode header = new ObjectMapper()
                    .createObjectNode()
                    .put(headerName, isHttpMaskingOn ? "[hidden]": response.getHeader(headerName));
            headers.add(header);
        }
    }
}
