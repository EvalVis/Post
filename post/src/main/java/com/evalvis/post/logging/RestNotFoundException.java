package com.evalvis.post.logging;

public class RestNotFoundException extends RuntimeException {
    public RestNotFoundException(String message) {
        super(message);
    }
}
