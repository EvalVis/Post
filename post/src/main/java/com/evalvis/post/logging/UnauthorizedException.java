package com.evalvis.post.logging;

public class UnauthorizedException extends RuntimeException {
    public UnauthorizedException(String message) {
        super(message);
    }
}
