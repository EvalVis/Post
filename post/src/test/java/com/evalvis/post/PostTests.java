package com.evalvis.post;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.evalvis.post.logging.RestNotFoundException;
import com.evalvis.security.JwtKey;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import org.assertj.core.api.Assertions;
import org.jsoup.safety.Safelist;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;

import javax.crypto.SecretKey;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

@ExtendWith({SnapshotExtension.class})
public class PostTests {
    private Expect expect;
    private final PostRepository repository;
    private final JwtKey jwtKey;
    private final PostController controller;
    private final PostMother mother;

    public PostTests() {
        this.repository = new FakePostRepository();
        this.jwtKey = new JwtKey() {
            @Override
            public SecretKey value() {
                return Keys.hmacShaKeyFor(
                        Decoders.BASE64.decode(
                                "b936cee86c9f87aa5d3c6f2e84cb5a4239a5fe50480a6ec66b70ab5b1f4ac6730c6c51542" +
                                        "1b327ec1d69402e53dfb49ad7381eb067b338fd7b0cb22247225d47"
                        )
                );
            }
        };
        this.controller = new PostController(repository, new FakeMinioStorage(), jwtKey, Safelist.basicWithImages());
        this.mother = new PostMother(this.controller, jwtKey);
    }

    @BeforeAll
    public static void setup() {
        SecurityContext securityContext = SecurityContextHolder.createEmptyContext();
        securityContext.setAuthentication(new FakeAuthentication("user@gmail.com", "password"));
        SecurityContextHolder.setContext(securityContext);
    }

    @Test
    void createsPost() {
        PostRepository.PostEntry post = mother.create(mother.request("u", "u@gmail.com"), "content");

        expect.toMatchSnapshot(jsonWithMaskedProperties(post, "postId", "datePosted"));
    }

    @Test
    void editsPost() {
        PostRepository.PostEntry post = mother.create(mother.request("a", "a@gmail.com"), "first");

        Post editedPost = mother.edit(new EditedPost(post.getPostId(), "second", "second"), 2);

        expect.toMatchSnapshot(jsonWithMaskedProperties(editedPost, "postId"));
    }

    @Test
    void findsEarlierPost() {
        PostRepository.PostEntry post = mother.create(mother.request("b", "b@gmail.com"), "earlier");

        mother.edit(new EditedPost(post.getPostId(), "newer", "newer"), 2);

        expect.toMatchSnapshot(
                controller.findByIdAndVersion(
                        post.getPostId(), 1, new FakeHttpServletRequest(), new FakeHttpServletResponse()
                )
        );
    }

    @Test
    void getsDateVersionMapping() {
        String id = mother.create(mother.request("c", "c@gmail.com"), "initial").getPostId();
        mother.edit(new EditedPost(id, "changed", "changed"), 2);
        mother.edit(new EditedPost(id, "changedAgain", "changedAgain"), 3);

        expect.toMatchSnapshot(
                jsonWithMaskedProperties(controller.getDateVersionMapping(id).getBody(), "postId", "datePosted")
        );
    }

    private <T> JsonNode jsonWithMaskedProperties(T object, String... properties) {
        try {
            JsonNode node = new ObjectMapper().valueToTree(object);
            node.forEach(n -> Arrays.stream(properties).forEach(property -> maskField(node, property)));
            return node;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static void maskField(JsonNode node, String property) {
        if (node.isObject()) {
            ObjectNode objectNode = (ObjectNode) node;
            if (objectNode.has(property)) {
                objectNode.put(property, "#hidden#");
            }
            objectNode.fields().forEachRemaining(entry -> maskField(entry.getValue(), property));
        } else if (node.isArray()) {
            for (JsonNode element : node) {
                maskField(element, property);
            }
        }
    }

    @Test
    void findsPost() {
        PostRepository.PostEntry initialPost = mother.create(mother.request("d", "d@gmail.com"), "content");

        Post foundPost = controller.findByIdAndVersion(
                initialPost.getPostId(), 1, new FakeHttpServletRequest(), new FakeHttpServletResponse()
        ).getBody();

        assertEquals(initialPost.getAuthorName(), foundPost.getAuthor());
        assertEquals(initialPost.getTitle(), foundPost.getTitle());
        assertEquals("content", foundPost.getContent());
    }

    @Test
    void findsAllPosts() {
        List<PostRepository.PostEntry> initialPosts = mother.createMultiple(mother.request("e", "e@gmail.com"));

        List<PostRepository.PostEntry> foundPosts = new ArrayList<>(controller.getAll().getBody());

        Assertions.assertThat(foundPosts)
                .extractingResultOf("toString")
                .containsExactlyInAnyOrderElementsOf(
                        initialPosts.stream().map(PostRepository.PostEntry::toString).toList()
                );
    }

    @Test
    void throwsExceptionIfSearchedPostDoesNotExist() {
        Exception e = assertThrows(
                RestNotFoundException.class,
                () -> controller.findByIdAndVersion(
                        "not-existing-id", 1, new FakeHttpServletRequest(), new FakeHttpServletResponse()
                )
        );
        assertEquals("Post with id: not-existing-id not found.", e.getMessage());
    }
}