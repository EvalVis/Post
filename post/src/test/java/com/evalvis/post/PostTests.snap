com.evalvis.post.PostTests.createsPost=[
  {
    "authorName": "author",
    "datePosted": "#hidden#",
    "postId": "#hidden#",
    "title": "title",
    "version": 1
  }
]


com.evalvis.post.PostTests.editsPost=[
  {
    "author": "author",
    "title": "second",
    "content": "second"
  }
]


com.evalvis.post.PostTests.findsEarlierPost=[
  {
    "body": {
      "author": "author",
      "content": "earlier",
      "title": "title"
    },
    "headers": { },
    "status": "OK"
  }
]


com.evalvis.post.PostTests.getsDateVersionMapping=[
  [
    {
      "authorName": "author",
      "datePosted": "#hidden#",
      "postId": "#hidden#",
      "title": "title",
      "version": 1
    },
    {
      "authorName": "author",
      "datePosted": "#hidden#",
      "postId": "#hidden#",
      "title": "changed",
      "version": 2
    },
    {
      "authorName": "author",
      "datePosted": "#hidden#",
      "postId": "#hidden#",
      "title": "changedAgain",
      "version": 3
    }
  ]
]